controllers/storagecluster/storagequota.go

```diff
@@ -0,0 +1,149 @@
+package storagecluster
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	quotav1 "github.com/openshift/api/quota/v1"
+	ocsv1 "github.com/openshift/ocs-operator/api/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ocsStorageQuota struct{}
+
+const (
+	clusterResourceQuotaKind       = "ClusterResourceQuota"
+	clusterResourceQuotaStorageKey = "ocs-clusterresourcequota"
+)
+
+// ensureCreated ensures that all ClusterResourceQuota resources exists with their Spec in
+// the desired state.
+func (obj *ocsStorageQuota) ensureCreated(r *StorageClusterReconciler, sc *ocsv1.StorageCluster) error {
+	for idx, opc := range sc.Spec.OverprovisionControl {
+		hardLimit := hardLimitOf(sc, &opc)
+		if hardLimit == nil {
+			continue
+		}
+		requestName := resourceRequestName(opc.StorageClassName)
+		storageQuota := &quotav1.ClusterResourceQuota{
+			TypeMeta:   metav1.TypeMeta{APIVersion: quotav1.SchemeGroupVersion.String(), Kind: clusterResourceQuotaKind},
+			ObjectMeta: metav1.ObjectMeta{Name: storageQuotaName(opc.StorageClassName, idx)},
+			Spec: quotav1.ClusterResourceQuotaSpec{
+				Selector: opc.Selector,
+				Quota: corev1.ResourceQuotaSpec{
+					Hard: corev1.ResourceList{requestName: *hardLimit},
+				},
+			},
+		}
+		currentQuota, err := r.QuotaV1.ClusterResourceQuotas().Get(context.TODO(), storageQuota.Name, metav1.GetOptions{})
```

Comment:
It will not work. OpenShift's ClusterResourceQuota has its own client interface. I do not know why they did it this way, but I tried to follow their conventions. See for example:

[](https://github.com/openshift/oc/blob/master/pkg/helpers/describe/describer.go#L1749)

[](https://github.com/openshift/oc/blob/master/pkg/cli/create/clusterquota.go#L140
)